contributors/design-proposals/sig-storage/volume-topology-scheduling.md

```diff
@@ -0,0 +1,580 @@
+For the alpha phase, the focus is on static provisioning of PVs to support
+persistent local storage.
+
+TODO: flow chart
+
+The proposed new workflow for volume binding is:
+1. Admin statically creates PVs and/or StorageClasses.
+2. User creates unbound PVC and there are no prebound PVs for it.
+3. **NEW:** PVC binding and provisioning is delayed until a pod is created that
+references it.
+4. User creates a pod that uses the PVC.
+5. Pod starts to get processed by the scheduler.
+6. **NEW:** A new predicate function, called MatchUnboundPVCs, will look at all of
+a Pod’s unbound PVCs, and try to find matching PVs for that node based on the
+PV topology.  If there are no matching PVs, then it checks if dynamic
+provisioning is possible for that node.
+7. **NEW:** The scheduler continues to evaluate priorities.  A new priority
+function, called PrioritizeUnboundPVCs, will get the PV matches per PVC per
+node, and compute a priority score based on various factors.
+8. **NEW:** After evaluating all the existing predicates and priorities, the
+scheduler will pick a node, and call a new assume function, AssumePVCs,
+passing in the Node.  The assume function will check if any binding or
+provisioning operations need to be done.  If so, it will update the PV cache to
+mark the PVs with the chosen PVCs.
+9. **NEW:** If PVC binding or provisioning is required, we do NOT AssumePod.
+Instead, a new bind function, BindPVCs, will be called asynchronously, passing
+in the selected node.  The bind function will prebind the PV to the PVC, or
+trigger dynamic provisioning, and then wait until the PVCs are bound
+successfully or encounter failure.  Then, it always sends the Pod through the
+scheduler again for reasons explained later.
+10. When a Pod makes a successful scheduler pass once all PVCs are bound, the
+scheduler assumes and binds the Pod to a Node.
+11. Kubelet starts the Pod.
+
+This new workflow will have the scheduler handle unbound PVCs by choosing PVs
+and prebinding them to the PVCs.  The PV controller completes the binding
+transaction, handling it as a prebound PV scenario.
+
+Prebound PVCs and PVs will still immediately be bound by the PV controller.
+
+One important point to note is that for the alpha phase, manual recovery is
+required in following error conditions:
+* A Pod has multiple PVCs, and only a subset of them successfully bind.
+* The scheduler chose a PV and prebound it, but the PVC could not be bound.
+
+In both of these scenarios, the primary cause for these errors is if a user
+prebinds other PVCs to a PV that the scheduler chose.  Some workarounds to
+avoid these error conditions are to:
+* Prebind the PV instead.
+* Separate out volumes that the user prebinds from the volumes that are
+available for the system to choose from by StorageClass.
+
+In addition, this new design will cause functional regression for the
+following use cases:
+* While a Pod is running, a user deletes the PVC and PV, and then recreates them.
```

Comment:
So you unmount the Pods storage and it doesn't die? What application can tolerate having a volume unmounted from under it without killing the underlying application.

Reply:
The volume does not get unmounted while the Pod is still running, even if the PVC/PV are deleted.  Unmount only happens when the Pod is terminated.  

However, if PVC is deleted, and reclaim policy is Delete, then PV controller will immediately try to delete the volume.  Luckily, nothing bad happens because it depends on the storage provider to prevent volume deletion while the disk is still attached to a node.  

But this is potentially a problem for local storage, which doesn't have that protection.  I need to think about if there's a reliable way to detect and prevent it (in a separate issue).